### PR TITLE
670 add bundle example for multi-gpu training

### LINKED ISSUE #670: Add multi-gpu example in the spleen segmentation bundle
**Is your feature request related to a problem? Please describe.**
Thanks for Ahmed's feedback, would be nice to have the example to show how to run training with multi-gpu in bundle.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #670 .

### Description
This PR added multi-gpu training example in the spleen bundle.

### Status
**Ready**

### Checks

- [ ] Notebook runs automatically `./runner [-p <regex_pattern>]`

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Hi @wyli @ericspod ,

I added 2 methods to support multi-gpu training in this PR:
(1) Newly defined config `train_multi_gpu.json` which contains all the logic.
(2) A much simpler config `multi_gpu_train.json` which only overrides the existing `train.json` config.

Which method do you prefer?

Thanks in advance.

**Comment by ericspod** [comment]
I think providing the override and an example of its usage would be a better idea. The override mechanism is quite powerful but I imagine for beginners it's not clear how it's meant to be used.

**Comment by Nic-Ma** [comment]
Depends on: https://github.com/Project-MONAI/MONAI/pull/4192.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @ericspod @wyli ,

Thanks for your review and comments.
I updated the PR to make it clearer with the **multiple configs** method instead of the "implicit" **args_file** method:
`--config_file [configs/train.json, configs/multi_gpu_train.json]`
It depends on this minor PR: https://github.com/Project-MONAI/MONAI/pull/4192.
Could you please help review it again?

Thanks in advance.

**Comment by Nic-Ma** [comment]
I will add the `evaluation` config in this PR soon.

Thanks.

**Comment by ericspod** [comment]
I approved and committed https://github.com/Project-MONAI/MONAI/pull/4192

**Comment by Nic-Ma** [comment]
> I approved and committed [Project-MONAI/MONAI#4192](https://github.com/Project-MONAI/MONAI/pull/4192)

Thanks @ericspod ,

@wyli @ericspod Then this PR is ready for review now.

Thanks.

File: modules/bundles/get_started.ipynb [hunk 1]
```diff
@@ -164,8 +164,6 @@
     "        \"$import os\",\n",
```
Comment (ericspod) [inline modules/bundles/get_started.ipynb:164]: <p><code></code><code></p><p>A MONAI bundle usually includes the stored weights of a model, TorchScript model, JSON files which include configs and metadata about the model, information for constructing training, inference, and post-processing transform sequences, plain-text description, legal information, and other data the model creator wishes to include.</p><p>For more information about MONAI bundles read the description: <a href="https://docs.monai.io/en/latest/bundle_intro.html" target="_blank">https://docs.monai.io/en/latest/bundle_intro.html</a>.</p><p>This notebook is a step-by-step tutorial to help get started to develop a bundle package, which contains a config file to construct the training pipeline and also has a <code>metadata.json</code> file to define the metadata information.</p><p>This notebook mainly contains the below sections:</p><ul><li>Define a training config with <code>JSON</code> or <code>YAML</code> format</li><li>Execute training based on bundle scripts and configs</li><li>Hybrid programming with config and python code</li></ul><p>You can find the usage examples of MONAI bundle key features and syntax in this tutorial, like:</p><ul><li>Instantiate a python object from a dictionary config with <code>_target_</code> indicating class or function name or module path.</li><li>Execute python expression from a string config with the <code>$</code> syntax.</li><li>Refer to other python object with the <code>@</code> syntax.</li><li>Require other independent config items to execute or instantiate first with the <code>_requires_</code> syntax.</li><li>Macro text replacement with the <code>%</code> syntax to simplify the config content.</li><li>Leverage the <code>_disabled_</code> syntax to tune or debug different components.</li><li>Override config content at runtime.</li><li>Hybrid programming with config and python code.</li></ul><p></code><code></code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Project-MONAI/tutorials/pull/673/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (ericspod) [inline modules/bundles/get_started.ipynb:164]: <p><code></code><code></p><p>Now let's start to define the config file for a regular training task. MONAI bundles support both <code>JSON</code> and <code>YAML</code> format, here we use <code>JSON</code> as the example.</p><p>According to the predefined syntax of MONAI bundle, <code>$</code> indicates an expression to evaluate, <code>@</code> refers to another object in the config content. For more details about the syntax in bundle config, please check: <a href="https://docs.monai.io/en/latest/config_syntax.html" target="_blank">https://docs.monai.io/en/latest/config_syntax.html</a>.</p><p>Please note that a MONAI bundle doesn't require any hard-coded logic in the config, so users can define the config content in any structure.</p><p>For the first step, import <code>os</code> and <code>glob</code> to use in the expressions (start with <code>$</code>), then define input / output environments and enable <code>cudnn.benchmark</code> for better performance.</p><p></code><code></code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Project-MONAI/tutorials/pull/673/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (ericspod) [inline modules/bundles/get_started.ipynb:164]: <p>I would have said the metadata file was required since it would include information we'd always want to have. </p><p> </p><p><code></code><code></p><p>Optionally, we can define a <code>metadata</code> file in the bundle, which contains the metadata information relating to the model, including what the shape and format of inputs and outputs are, what the meaning of the outputs are, what type of model is present, and other information. The structure is a dictionary containing a defined set of keys with additional user-specified keys.</p><p></code><code></code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Project-MONAI/tutorials/pull/673/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (ericspod) [inline modules/bundles/get_started.ipynb:164]: <pre class="ql-syntax" spellcheck="false">A MONAI bundle supports flexible customized logic, there are several ways to achieve this:

If defining own components like transform, loss, trainer, etc. in a python file, just use its module path in _target_ within the config file.
Parse the config in your own python program and do lazy instantiation with customized logic.

Here we show an example to parse the config in python code and execute the training.
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Project-MONAI/tutorials/pull/673/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

**Comment by ericspod** [comment]
I tried reviewing through ReviewNB, I'm not sure it's as coherent as I'd hoped.

When you have a structure like:
```json
"train":{
...some defs...
}
```
The "train" value is interpreted as a `dict` by the parser correct? So in places where you have "train#loss" or something like it that sets that entry in that dictionary. I think that needs to be explained in a bit more detail to make sense to the reader and to understand what to use it for.

**Review by ericspod — COMMENTED** [review]
I had commented on a few changes to text and a question on references to members of dictionaries (I think) like "validate#postprocessing" that we should explain at little more.

File: modules/bundles/get_started.ipynb (hunk 1 above)
Comment (Nic-Ma) [reply modules/bundles/get_started.ipynb:163]: Sure, updated in the latest commit.

Thanks.

Comment (Nic-Ma) [reply modules/bundles/get_started.ipynb:163]: Sure, updated in the latest commit.

Thanks.

Comment (Nic-Ma) [reply modules/bundles/get_started.ipynb:163]: Sure, updated in the latest commit.

Thanks.

Comment (Nic-Ma) [reply modules/bundles/get_started.ipynb:163]: Sure, updated in the latest commit.

Thanks.

**Comment by Nic-Ma** [comment]
> I tried reviewing through ReviewNB, I'm not sure it's as coherent as I'd hoped.
> 
> When you have a structure like:
> 
> ```json
> "train":{
> ...some defs...
> }
> ```
> 
> The "train" value is interpreted as a `dict` by the parser correct? So in places where you have "train#loss" or something like it that sets that entry in that dictionary. I think that needs to be explained in a bit more detail to make sense to the reader and to understand what to use it for.

Good point, I added more description for it in the `dataset` and `dataloader` section.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

Thanks for your review and comments, I updated the PR according to them.
Could you please help review again?

Thanks in advance.

**Review by ericspod — APPROVED** [review]
Looks good!
